laplace/baselaplace.py

```diff
@@ -1568,8 +1567,1025 @@ def sample(self, n_samples=100, generator=None):
+
+    Note that for `likelihood='classification'`, we approximate \( L_{NN} \\) with a diagonal matrix
+    ( \\( L_{NN} \\) is a block-diagonal matrix, where blocks represent Hessians of per-data-point log-likelihood w.r.t.
+    neural network output \\( f \\), See Appendix [A.2.1](https://arxiv.org/abs/2008.08400) for exact definition). We
+    resort to such an approximation because of the (possible) errors found in Laplace approximation for
+    multiclass GP classification in Chapter 3.5 of [R&W 2006 GP book](http://www.gaussianprocess.org/gpml/),
+    see the question
+    [here](https://stats.stackexchange.com/questions/555183/gaussian-processes-multi-class-laplace-approximation)
+    for more details. Alternatively, one could also resort to *one-vs-one* or *one-vs-rest* implementations
+    for multiclass classification, however, that is not (yet) supported here.
+
+    Parameters
+    ----------
+    M : int
+        number of data points for Subset-of-Data (SOD) approximate GP inference.
+        By default (`M=None`), all data points from train dataset are used
+    diagonal_kernel : bool
+        GP kernel here is product of Jacobians, which results in a \\( C \\times C\\) matrix where \\(C\\) is the output
+        dimension. If `diagonal_kernel=True`, only a diagonal of a GP kernel is used. This is (somewhat) equivalent to
+        assuming independent GPs across output channels.
+
+    See `BaseLaplace` class for the full interface.
+    """
+
+    # key to map to correct subclass of BaseLaplace, (subset of weights, Hessian structure)
+    _key = ('all', 'gp')
+
+    def __init__(
+        self,
+        model,
+        likelihood,
+        M=None,
+        sigma_noise=1.0,
+        prior_precision=1.0,
+        prior_mean=0.0,
+        temperature=1.0,
+        backend=BackPackGGN,
+        backend_kwargs=None,
+        enable_backprop=False,
+        diagonal_kernel=False,
+        seed=0,
+    ):
+        assert backend in [BackPackGGN, AsdlGGN]
+        self._check_prior_precision(prior_precision)
+        super().__init__(
+            model,
+            likelihood,
+            sigma_noise,
+            prior_precision,
+            prior_mean,
+            temperature,
+            backend,
+            backend_kwargs,
+        )
+        self.enable_backprop = enable_backprop
+
+        self.M = M
+        self.diagonal_kernel = diagonal_kernel
+        self.seed = seed
+
+        self.K_MM = None
+        self.Sigma_inv = None  # (K_{MM} + L_MM_inv)^{-1}
+        self.train_loader = (
+            None  # needed in functional variance and marginal log likelihood
+        )
+        self.batch_size = None
+        self._prior_factor_sod = None
+        self.mu = None  # mean in the scatter term of the log marginal likelihood
+        self.L = None
+
+        # MAP estimate of NN parameters (used in regression marginal likelihood)
+        self.map_estimate = parameters_to_vector(self.model.parameters()).detach()
```

Comment:
I believe `BaseLaplace` already store the same thing in `self.mean`. If possible, avoid duplicates to keep memory requirements low

Reply:
Actually yes! Changed that!